### PR TITLE
Do not open class/loadout menu if you are spectating

### LINKED ISSUE #385: [BUG] Spectators shown the class menu on round start
### Build Info

20240620_f670f82 (master-f670f828687a22886a4d6b89fc09fcaea9b23e70)

### Description

Players spectator the game (not in Jinrai or NSF) are shown the class menu on the start of each round. This is despite spectators aren't playing and don't have class or weapon selection.

### To Reproduce

1. Open game -> Create Server -> Start
2. Pick spectator
3. Go to console and create bots to start a match:
   ```
   sv_cheats 1
   mp_neo_warmup_round_time 0
   bot_add;bot_add;bot_add;bot_add;
   ```
4. Wait till round starts

### Expected behavior

There's no menu automatically shown for the spectator on round start

### Actual behavior

There's a class menu automatically shown for the spectator on round start

### Operating System

- [X] Windows
- [X] Linux

### Version/Distro

Arch Linux; Windows 10

### Machine's CPU

_No response_

### Machine's GPU

_No response_

### GPU's driver

_No response_

### Build's compiler

_No response_

### Additional context and Screenshots

The class menu automatically pops up even as a separator:
2024-06-20T211555_screen

### PR DESCRIPTION
**Pull request by VantoNortim**
## Description
Do not open the Class or Loadout menu while being in team Spectate/Unassigned

## Toolchain
- Windows MSVC VS2022

## Linked Issues
- fixes #385

### CODE REVIEW

**Comment by nullsystem** [comment]
@VantoNortim btw this branch is very behind the current master branch

**Review by nullsystem — APPROVED** [review]
Looks fine after rebasing it to latest master

File: mp/src/game/client/neo/c_neo_player.cpp [hunk 1]
```diff
@@ -111,6 +111,12 @@ class NeoLoadoutMenu_Cb : public ICommandCallback
 public:
 	virtual void CommandCallback(const CCommand& command)
 	{
+		auto team = GetLocalPlayerTeam();
+
+		if (team == TEAM_UNASSIGNED || team == TEAM_SPECTATOR) {
+			return;
+		}
```
Comment (Rainyan) [inline mp/src/game/client/neo/c_neo_player.cpp:118]: Nit: from the shareddefs header, using `if (GetLocalPlayerTeam() < FIRST_GAME_TEAM)` would be a way to express this more tersely.

**Review by Rainyan — CHANGES_REQUESTED** [review]
This prevents the  player from picking their class and loadout when they are first joining a playing team from spectator, because the associated `jointeam` command has not been processed yet.

The simplest solution is probably to just assign the client's team locally before these callbacks are invoked, for example:

```diff
diff --git a/mp/src/game/client/neo/game_controls/neo_teammenu.cpp b/mp/src/game/client/neo/game_controls/neo_teammenu.cpp
index 8658be7..1d095b0 100644
--- a/mp/src/game/client/neo/game_controls/neo_teammenu.cpp
+++ b/mp/src/game/client/neo/game_controls/neo_teammenu.cpp
@@ -177,6 +177,7 @@ void CNeoTeamMenu::OnCommand(const char *command)
 		const int nsfNumPlayers = (pNsf != NULL ? pNsf->Get_Number_Players() : 0);
 		int randomTeam = jinNumPlayers > nsfNumPlayers ? TEAM_NSF : (nsfNumPlayers > jinNumPlayers ? TEAM_JINRAI : RandomInt(TEAM_JINRAI, TEAM_NSF));
 		V_sprintf_safe(commandBuffer, "jointeam %i", randomTeam);
+		C_NEO_Player::GetLocalNEOPlayer()->m_iTeamNum = randomTeam;
 		ChangeMenu("classmenu");
 		engine->ClientCmd(commandBuffer);
 		return;
@@ -197,7 +198,25 @@ void CNeoTeamMenu::OnCommand(const char *command)
 	}
 
 	if (Q_stristr(commandBuffer, "jointeam") != 0) // Note using stristr, not strcmp. Equates to true when jointeam in commandBuffer
-	{ // joining jinrai or nsf
+	{ // joining something other than TEAM_SPECTATOR or TEAM_NONE
+		CUtlStringList strs;
+		V_SplitString(commandBuffer, "jointeam ", strs);
+		if (strs.Count() != 0)
+		{
+			char* szNum = strs[0];
+			const int len = V_strlen(szNum);
+			for (int i = 0; i < len; ++i)
+			{
+				if (!V_isdigit(szNum[i]))
+				{
+					szNum[i] = '\0';
+					break;
+				}
+			}
+			const int team = V_atoi(szNum);
+			C_NEO_Player::GetLocalNEOPlayer()->m_iTeamNum = team;
+		}
+
 		ChangeMenu("classmenu");
 		engine->ClientCmd(commandBuffer);
 		return;
```

But maybe double check that C string manipulation stuff :smile:

The `m_iTeamNum` is networked, so any discrepancy should get overwritten by the server once it has processed the team request.

**Comment by VantoNortim** [comment]
> This prevents the player from picking their class and loadout when they are first joining a playing team from spectator, because the associated `jointeam` command has not been processed yet.
> 
> The simplest solution is probably to just assign the client's team locally before these callbacks are invoked, for example:
> 
> ```diff
> diff --git a/mp/src/game/client/neo/game_controls/neo_teammenu.cpp b/mp/src/game/client/neo/game_controls/neo_teammenu.cpp
> index 8658be7..1d095b0 100644
> --- a/mp/src/game/client/neo/game_controls/neo_teammenu.cpp
> +++ b/mp/src/game/client/neo/game_controls/neo_teammenu.cpp
> @@ -177,6 +177,7 @@ void CNeoTeamMenu::OnCommand(const char *command)
>  		const int nsfNumPlayers = (pNsf != NULL ? pNsf->Get_Number_Players() : 0);
>  		int randomTeam = jinNumPlayers > nsfNumPlayers ? TEAM_NSF : (nsfNumPlayers > jinNumPlayers ? TEAM_JINRAI : RandomInt(TEAM_JINRAI, TEAM_NSF));
>  		V_sprintf_safe(commandBuffer, "jointeam %i", randomTeam);
> +		C_NEO_Player::GetLocalNEOPlayer()->m_iTeamNum = randomTeam;
>  		ChangeMenu("classmenu");
>  		engine->ClientCmd(commandBuffer);
>  		return;
> @@ -197,7 +198,25 @@ void CNeoTeamMenu::OnCommand(const char *command)
>  	}
>  
>  	if (Q_stristr(commandBuffer, "jointeam") != 0) // Note using stristr, not strcmp. Equates to true when jointeam in commandBuffer
> -	{ // joining jinrai or nsf
> +	{ // joining something other than TEAM_SPECTATOR or TEAM_NONE
> +		CUtlStringList strs;
> +		V_SplitString(commandBuffer, "jointeam ", strs);
> +		if (strs.Count() != 0)
> +		{
> +			char* szNum = strs[0];
> +			const int len = V_strlen(szNum);
> +			for (int i = 0; i < len; ++i)
> +			{
> +				if (!V_isdigit(szNum[i]))
> +				{
> +					szNum[i] = '\0';
> +					break;
> +				}
> +			}
> +			const int team = V_atoi(szNum);
> +			C_NEO_Player::GetLocalNEOPlayer()->m_iTeamNum = team;
> +		}
> +
>  		ChangeMenu("classmenu");
>  		engine->ClientCmd(commandBuffer);
>  		return;
> ```
> 
> But maybe double check that C string manipulation stuff 😄
> 
> The `m_iTeamNum` is networked, so any discrepancy should get overwritten by the server once it has processed the team request, so there's no need to do super strict validation here.

I found a better way to do this, by just forcing the team change to execute first. This way it just works.

**Comment by Rainyan** [comment]
I still can't see the class and loadout menus appear. Breakpointing on the `engine->ExecuteClientCmd`, and then continuing execution makes them appear, but just running the game I get:

* Pick team
* No further menus appear to choose from
* You just sort of spawn as assault + MPN

**Comment by Rainyan** [comment]
[class_loadout.webm](https://github.com/NeotokyoRebuild/neo/assets/6595066/ca3a7533-77de-4eee-9338-db45ebfa67ed)

Well, it's still doing this thing for me where I can't choose a class or loadout. 🤷‍♂️

**Comment by AdamTadeusz** [comment]
> class_loadout.webm
> 
> Well, it's still doing this thing for me where I can't choose a class or loadout. 🤷‍♂️ Perhaps someone else should test this on their machine to see if it's an issue on my end.

Is this on a dedicated server? Just tested on a listen server and having no trouble selecting team and loadout

**Comment by Rainyan** [comment]
> Is this on a dedicated server?

This is listen server. I've tried with a full rebuild and it still occurs. But I guess it's verified an issue on my end then, whatever it is.

**Comment by AdamTadeusz** [comment]
I noticed on dedicated the class selection menu doesn;t open up after pickign a team like on a listen server, but I can still use f2 to open it and select that way. Now I gotta check master to see whether the same behaviour is on there

**Comment by AdamTadeusz** [comment]
> > Is this on a dedicated server?
> 
> This is listen server. I've tried with a full rebuild and it still occurs. But I guess it's verified an issue on my end then, whatever it is.

So statistically speaking 33% of the playerbase will not be able to select their class after they join a server?

**Comment by VantoNortim** [comment]
I am going to need more time to look into this then, I would have thought there is an event client side when the player switches teams. Maybe I should clean up opening those elements in general there are so many bools in the player class related to them. But surely we could just check if any of them are visible and so on.

**Review by AdamTadeusz — CHANGES_REQUESTED** [review]
Yeah I wouldn't pull this in right now since when joining a team on a dedicated server when you are initially a spectator the class menu doesn't open up.

Probably something to do with latency where on low latency your client picks up that you successfully joined a team and opens the menu in time, but with higher latency the client thinks you're still a spectator when it checks whether it should open the menu so doesn't show the menu the first time around

While using the keybind still seems to work for me I think most people might not be aware of the bind and thus can not pick their class

File: mp/src/game/client/neo/c_neo_player.cpp [hunk 2]
```diff
@@ -284,6 +290,12 @@ class NeoClassMenu_Cb : public ICommandCallback
 public:
 	virtual void CommandCallback(const CCommand& command)
 	{
+		auto team = GetLocalPlayerTeam();
+
+		if (team == TEAM_UNASSIGNED || team == TEAM_SPECTATOR) {
```
Comment (AdamTadeusz) [inline mp/src/game/client/neo/c_neo_player.cpp:295]: Can probably use FIRST_GAME_TEAM here same as in neoloadoutmenu

**Comment by AdamTadeusz** [comment]
I noticed you were setting the value of m_iTeam locally before making a request to change the local player's team to that team via a command, so I changed how the function to open the class and loadout menus check what team the local player is on to this and now the problems I mentioned before are gone 
image

**Comment by VantoNortim** [comment]
> I noticed you were setting the value of m_iTeam locally before making a request to change the local player's team to that team via a command, so I changed how the function to open the class and loadout menus check what team the local player is on to this and now the problems I mentioned before are gone image
> 
> @Rainyan Could you do the same things I did in the two functions on line 216 and 293 and then see if the problem is resolved?
> 
> (edit) this is in c_neo_player
> 
> (edit 2) doing it this way isn't actually any different from the original, I have no clue why this fixed it for me, probably still a race condition thing

Yeah I think I will have to hook on the team change event sent by server and once that is ran I will show the class menu.

**Pull request closed without merging**